### DISCUSSION TITLE
Environment variables

### DISCUSSION THREAD

**Discussion by AbanoubZaki** [codemagic-ci-cd/codemagic-docs#1982] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [x] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

If i want to use SHA-256 certificates in the dart code but i don't want them to be on the repository.

### Expected results

How to create environment variables and access it inside the dart code?

### Actual results

I don't know how to access environment variables inside the dart code.

**Comment by kulakarla** [comment, marked as answer]
Hello @AbanoubZaki 

Dart has different ways of using environment variables.

One way to do it would be like this:

- Add your environment variables to Codemagic (for example, CM_TESTENV1, CM_TESTENV2)
- When building the flutter app, add `--dart-define` to your scripts. E.g `flutter build apk --dart-define=API_KEY=$CM_TESTENV1 --dart-define=SOMEOTHERVAR=$CM_TESTENV2`
- You can now access these variables in your code with `String.fromEnvironment` method. Example: `String.fromEnvironment('API_KEY')`


Refer to [this article](https://www.sandromaglione.com/techblog/how-to-use-environmental-variables-in-flutter), which provides multiple ways of dealing with environment variables in Flutter. The aforementioned described method is one of the proposed ways in the article, a relatively easy one.

Let me know if manage to get everything configured to your needs. If issues occur, let me know.

**Reply by ntfnd404** [reply]
How to specify "--dart-define" when building in codemagic? i use Workflow Editor

**Reply by truszko1** [reply]
@ntfnd404 did you ever figure this out? I'm having the same confusion..

**Reply by CedricBaijot** [reply]
You can add it under Build>Build arguments. With my 3 environment variables BACKEND_URL, SUPABASE_ANNON_KEY and SUPABASE_URL it looks like this:
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/140388494/60a0fc27-fc3f-4761-9ee8-388f253e7aa5)
`--build-name=1.0.$PROJECT_BUILD_NUMBER --build-number=$PROJECT_BUILD_NUMBER --dart-define=BACKEND_URL=$BACKEND_URL --dart-define=SUPABASE_ANNON_KEY=$SUPABASE_ANNON_KEY --dart-define=SUPABASE_URL=$SUPABASE_URL`
